Project-Aurora/Project-Aurora/Devices/Ducky/DuckyDevice.cs

```diff
@@ -0,0 +1,360 @@
+            colourMessage[Packet(9) + 1] = 0x51;
+            colourMessage[Packet(9) + 2] = 0x28;
+            colourMessage[Packet(9) + 5] = 0xFF;
+
+            shine7Keyboard = DeviceList.Local.GetHidDevices(0x04D9, 0x0348).SingleOrDefault(HidDevice => HidDevice.GetMaxInputReportLength() == 65);
+            try
+            {
+                isInitialized = shine7Keyboard.TryOpen(out packetStream);
+                //This uses a monstrous 501 packets to initialize the keyboard in to letting the LEDs be controlled over USB HID.
+                foreach (byte[] controlPacket in duckyAPI.getControlCommand("Shine_7_Takeover"))
+                {
+                    packetStream.Write(controlPacket);
+                }
+            }
+            catch
+            {
+                isInitialized = false;
+            }
+            
+            return isInitialized;
+        }
+
+        public void Shutdown()
+        {
+            //This one is a little smaller, 81 packets. This tells the keyboard to no longer allow USB HID control of the LEDs.
+            //You can tell both the takeover and release work because the keyboard will flash the same as switching to profile 1. (The same lights when you push FN + 1)
+            foreach (byte[] controlPacket in duckyAPI.getControlCommand("Shine_7_Release"))
+            {
+                try
+                {
+                    packetStream.Write(controlPacket);
+                }
+                catch { }
+            }
+            packetStream.Dispose();
+            packetStream.Close();
+            isInitialized = false;
+        }
+
+        public void Reset()
+        {
+            this.Shutdown();
+            this.Initialize();
+        }
+
+        public bool Reconnect()
+        {
+            throw new NotImplementedException();
+        }
+
+        public bool IsInitialized()
+        {
+            return isInitialized;
+        }
+
+        public bool IsConnected()
+        {
+            throw new NotImplementedException();
+        }
+
+        public bool IsKeyboardConnected()
+        {
+            return isInitialized;
+        }
+
+        public bool IsPeripheralConnected()
+        {
+            return isInitialized;
+        }
+
+        public bool UpdateDevice(Dictionary<DeviceKeys, Color> keyColors, DoWorkEventArgs e, bool forced = false)
+        {
+            foreach (KeyValuePair<DeviceKeys, Color> kc in keyColors)
+            {
+                //This keyboard doesn't take alpha (transparency) values, so we do this:
+                processedColor = ColorUtils.CorrectWithAlpha(kc.Value);
+
+                //This if statement grabs the packet offset from the key that Aurora wants to set, using packetOffsetMap.
+                //It also checks whether the key exists in the Dictionary, and if not, doesn't try and set the key colour.
+                if(!packetOffsetMap.TryGetValue(kc.Key, out currentOffset)){
```

Comment:
I don't think it's clear that `currentOffset` just contains which packet the key is supposed to be sent in, and the index where the colors should go. Maybe storing this as a named tuple instead of a random byte array would be better? would remove the need for the `PACKET_NUM` and `OFFSET_NUM` constants as well.